### PR TITLE
[MAINT] Add return type annotations to nilearn.image.resampling

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Related to #3568

Use of AI tools:
- [x] LLM tools were used to generate at least part of the content of this pull-request.
- [ ] No LLM tools were used to generate any part of the content of this pull-request.

Changes proposed in this pull request:

- Add return type annotations to `coord_transform`, `reorder_img`, `resample_img`, and `resample_to_img` in `nilearn.image.resampling`.
- Adjust a couple of internal type-inference edge cases (float64 tuple reassignment, slices tuple) so `mypy` passes on the now partially-typed functions.
- Add a changelog entry.

### CODE REVIEW

File: nilearn/image/tests/test_resampling.py [hunk 1]
```diff
@@ -104,6 +104,33 @@ def test_identity_resample_array_like(
...
 
 
+@pytest.mark.parametrize(
+    "target_affine",
+    [
+        _affine_eye().tolist(),
+        tuple(_affine_eye().tolist()),
+    ],
+)
+@pytest.mark.ai_generated
```
Comment (Remi-Gau) [inline nilearn/image/tests/test_resampling.py]:
```suggestion
```

File: doc/changes/latest.rst [hunk 2]
```diff
@@ -22,6 +22,8 @@ NEW
 Fixes
 -----
 
+- :bdg-dark:`Code` Fix :func:`~image.resample_img` raising an ``AttributeError`` instead of resampling correctly when ``target_affine`` is passed as a :obj:`list` or :obj:`tuple` together with ``target_shape`` (:gh:`6408` by `Rémi Gau`_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:25]: discovered while adding type annotations
